runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -5253,15 +5265,29 @@ class INTERPRETER_CLASS
 		_currentThread->callOutCount += 1;
 #endif /* JAVA_SPEC_VERSION >= 19 */
 		updateVMStruct(REGISTER_ARGS);
+#if JAVA_SPEC_VERSION >= 21
+		/* Remove VMaccess in the trivial downcall to benefit the JIT for better performance. */
+		if (!_currentThread->isTrivialOption) {
+			VM_VMAccess::inlineExitVMToJNI(_currentThread);
+		}
+#else /* JAVA_SPEC_VERSION >= 21 */
 		VM_VMAccess::inlineExitVMToJNI(_currentThread);
+#endif /* JAVA_SPEC_VERSION >= 21 */
 		VM_VMHelpers::beforeJNICall(_currentThread);
 #if FFI_NATIVE_RAW_API
 		ffiCallWithSetJmpForUpcall(_currentThread, cif, function, returnStorage, values, values_raw);
 #else /* FFI_NATIVE_RAW_API */
 		ffiCallWithSetJmpForUpcall(_currentThread, cif, function, returnStorage, values);
 #endif /* FFI_NATIVE_RAW_API */
 		VM_VMHelpers::afterJNICall(_currentThread);
+#if JAVA_SPEC_VERSION >= 21
+		/* Remove VMaccess in the trivial downcall to benefit the JIT for better performance. */
+		if (!_currentThread->isTrivialOption) {
```

Comment:
us `isTrivial` limited to JDK natives? as in no user natives can be used with this?

Reply:
> is isTrivial limited to JDK natives? as in no user natives can be used with this?

`isTrivial` is a public API specified in JDK21 at https://download.java.net/java/early_access/jdk21/docs/api/java.base/java/lang/foreign/Linker.Option.html#isTrivial(), in which case users are able to offer such option in their java application to trigger the behavior in downcall (native). That being said, there is no restriction for the option when it comes to JDK natives or user natives in downcall.